packages/grid/x-data-grid-premium/src/hooks/features/cellSelection/useGridCellSelection.ts

```diff
@@ -486,7 +488,9 @@ export const useGridCellSelection = (
         newClasses.push(gridClasses['cell--rangeTop']);
       }
 
-      if (rowIndex < visibleRows.range.lastRowIndex) {
+      // https://github.com/mui/mui-x/issues/10777
+      const rowIndexRelativeToRange = rowIndex + paginationModel.page * paginationModel.pageSize;
```

Comment:
Alternatively, we could compute this by:
```suggestion
      const rowIndexRelativeToRange = rowIndex + visibleRows.range.firstRowIndex;
```
Not sure how reliable this will be though.